codecs_derive/src/aper/seq.rs

```diff
@@ -21,15 +21,22 @@ pub(super) fn generate_aper_decode_for_asn_sequence(
     if fld_tokens.is_err() {
         return fld_tokens.err().unwrap().to_compile_error().into();
     }
-    let fld_tokens = fld_tokens.unwrap();
+    let (fld_decode_tokens, fld_encode_tokens) = fld_tokens.unwrap();
 
     let tokens = quote! {
         impl asn1_codecs::aper::AperCodec for #name {
             type Output = Self;
 
             fn decode(data: &mut asn1_codecs::aper::AperCodecData) -> Result<Self::Output, asn1_codecs::aper::AperCodecError> {
                 let (bitmap, _extensions_present) = asn1_codecs::aper::decode::decode_sequence_header(data, #ext, #opt_count)?;
-                Ok(Self{#(#fld_tokens)*})
+                Ok(Self{#(#fld_decode_tokens)*})
+            }
+
+            fn encode(&self, data: &mut asn1_codecs::aper::AperCodecData) -> Result<(), asn1_codecs::aper::AperCodecError> {
+                let mut bitmap = bitvec::bitvec![bitvec::prelude::Msb0, u8; 0; #opt_count];
+                asn1_codecs::aper::encode::encode_sequence_header(data, #ext, &bitmap, false)?;
```

Comment:
@nplrkn If you look at this line, the way I am trying to implement `encode` for sequence actually doesn't work well with current implementation of `encode_sequence_header`. It is not wrong per se, but it doesn't fit with the `codecs_derive`s implementation. 

So I am curious how you are using it. Because here's the issue - 
1. the sequence header should be populated separately based on `OPTIONAL` values being encoded and 'finally' be pushed ahead of the actual "bitarray". That does not work with current implementation because a) Either we change the signature to return a `slice` of the 'internal bit buffer' of `AperCodec` or b) Change the implementation to account for `prepending` to the internal buffer. 
2. Or I am missing something and there's a way to handle. How are you handling `encode` for sequence?

Reply:
Here's an example of how I'm doing it.

```
    fn encode_inner(&self, data: &mut AperCodecData) -> Result<(), AperCodecError> {
        let mut optionals = BitVec::new();
        optionals.push(self.additional_dl_qos_flow_per_tnl_information.is_some());
        optionals.push(self.security_result.is_some());
        optionals.push(self.qos_flow_failed_to_setup_list.is_some());
        optionals.push(false);

        aper::encode::encode_sequence_header(data, true, &optionals, false)?;
        self.dl_qos_flow_per_tnl_information.encode(data)?;
        if let Some(x) = &self.additional_dl_qos_flow_per_tnl_information {
            x.encode(data)?;
        }
        if let Some(x) = &self.security_result {
            x.encode(data)?;
        }
        if let Some(x) = &self.qos_flow_failed_to_setup_list {
            x.encode(data)?;
        }

        Ok(())
    }
```

Reply:
I see.. thanks! The reason I asked was that - I wanted to change the `encode_seq_header` function a bit. 

It looks like you are generating `encode` code for all of the functions by hand. That is just too much of code to be written by hand. My idea of encode was something like this - 

1. generate the `bitmap` equal in length to the # of optionals (this we get from the `struct` level attribute.
2. Have actual encode functions
3. Finally push the sequence header (thus will need to add a code to shift the `AperCodecData` buffer behind the `optionals` bitmap. I guess it can be accommodated in your code. do you think this change is okay for you right now?

Also, what I believe is the `codec` code would mostly be written for being used with `codec_derive` macros. Writing that code by hand is just too much. 

Your thoughts?

Reply:
In my case I am autogenerating the code directly from the ASN.1 (not writing it by hand, also not using derive macros).

Not sure there is any need to change `encode_sequence_header()`, as I think this might just involve adding new methods to jump back and forth in the buffer.

```
let sequence_header_bits_required = ...;

// leave some space for the sequence header and save its location to come back to later
let sequence_header_offset = data.advance(sequence_header_bits_required);

// encode fields and build up optionals bitfield

// go back and add the sequence header now we've got the bitfield
let end = data.seek(sequence_header_offset);
encode_sequence_header(data, ...);
data.seek(end);
```

Anyway, even if there are breaking changes to the API, I expect I'll be fine changing my autogenerator to accommodate them.

Reply:
No: I didn't want to bring about breaking change as much as possible, hence I requested suggestions, I guess the approach you mentioned above can be followed. Except the `offset` may needed to be called `offset` rather than `decode_offset` as it is called.